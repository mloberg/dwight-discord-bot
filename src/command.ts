import { SlashCommandBuilder, SlashCommandIntegerOption } from '@discordjs/builders';
import { CommandInteraction } from 'discord.js';

type Handler = (command: CommandInteraction) => Promise<void>;

export class CommandBuilder extends SlashCommandBuilder {
    constructor(readonly handler: Handler) {
        super();
    }
}

export const rollOption = (dice: string): SlashCommandIntegerOption =>
    new SlashCommandIntegerOption().setName('roll').setDescription(`Value of a ${dice} roll.`);
