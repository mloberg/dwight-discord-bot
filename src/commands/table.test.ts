import { Message } from 'discord.js';

import { FriendlyError } from '../error';
import command from './table';

jest.mock('discord.js');
jest.mock('../data/table', () => ({
    a: new Array(100).fill('foo'),
    b: [...new Array(98).fill('bar'), 'test', 'foobar'],
    c: new Array(100).fill(() => 'baz'),
}));

describe('_table configuration', () => {
    it('should have basic command infomation', () => {
        expect(command.name).toEqual('table');
        expect(command.description).toEqual('Roll on the magic item tables');
        expect(command.usage).toEqual('<table> [d100]');
        expect(command.alias).toEqual([]);
    });
});

describe('_table', () => {
    it('returns a value from the given table', async () => {
        const message = new Message({} as never, {} as never);
        await command.run(message, { command: 'table', args: ['a'], match: [], groups: {} });

        expect(message.delete).toBeCalledTimes(1);
        expect(message.reply).toBeCalledTimes(1);
        expect(message.reply).toHaveBeenCalledWith('foo');
    });

    it('returns a value for a dice roll', async () => {
        const message = new Message({} as never, {} as never);
        await command.run(message, { command: 'table', args: ['b', '99'], match: [], groups: {} });

        expect(message.delete).toBeCalledTimes(1);
        expect(message.reply).toBeCalledTimes(1);
        expect(message.reply).toHaveBeenCalledWith('test');
    });

    it('returns a resolved value', async () => {
        const message = new Message({} as never, {} as never);
        await command.run(message, { command: 'table', args: ['C'], match: [], groups: {} });

        expect(message.delete).toBeCalledTimes(1);
        expect(message.reply).toBeCalledTimes(1);
        expect(message.reply).toHaveBeenCalledWith('baz');
    });

    it('throws an error if invalid table given', async () => {
        const message = new Message({} as never, {} as never);

        try {
            await command.run(message, { command: 'table', args: [], match: [], groups: {} });
            fail('expected error to be thrown');
        } catch (err) {
            expect(err).toBeInstanceOf(FriendlyError);
            expect(err.message).toEqual('Unknown table "-".');
        }
    });
});
