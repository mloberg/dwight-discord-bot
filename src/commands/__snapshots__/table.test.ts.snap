// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/table is a slash command 1`] = `
Object {
  "description": "Roll on the magic item tables",
  "name": "table",
  "options": Array [
    Object {
      "choices": Array [
        Object {
          "name": "A",
          "value": "a",
        },
        Object {
          "name": "B",
          "value": "b",
        },
        Object {
          "name": "C",
          "value": "c",
        },
      ],
      "description": "Magic item table to roll on",
      "name": "table",
      "required": true,
      "type": 3,
    },
    Object {
      "choices": undefined,
      "description": "Value of a d100 roll.",
      "name": "roll",
      "required": false,
      "type": 4,
    },
  ],
}
`;
